[Runus01] iP

Runus01 opened the pull request:
# **Elsa: Your Task Companion**
> “Let’s focus on getting things done, not holding them in our heads!” – Elsa ([source](https://i.pinimg.com/originals/77/f5/6b/77f56bc55cdbc39e9971b5bd64989f3d.jpg))

Elsa's here to take the weight off your mind. Managing your tasks has never been easier. She's:

- Conversational and friendly
- Easy to get to know
- Lightning-fast ⚡ (Seriously, she’s ~~speedy~~ _super_ speedy!)

Getting started is simple:

1. Download Elsa from [here](https://github.com/Runus01/ip/releases/download/A-Jar/Elsa.jar).
2. Double-click to open and say hi!
3. Add your tasks – and voilà, Elsa's on it!

And guess what? Elsa is completely **FREE**!

Features Elsa offers:

- [x] Managing your tasks effortlessly
- [ ] Deadline tracking: Elsa's planning it
- [ ] Reminders: For when we need that gentle nudge (coming soon)
---
Are you a Java developer? Elsa can even help you hone your skills while managing your tasks. Check out the main method for starting her up:
```java
public class Elsa {
    public static void main(String[] args) {
        new Elsa().run();
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

Runus01 committed:
Level 0 - 1st update

Runus01 committed:
Added a method that prints a horizontal line and a method to say goodbye

Runus01 committed:
Added a method to process userInput

Runus01 committed:
Added the ability to store text entered by the user and display it back when requested

Runus01 committed:
Add the ability to mark tasks as done or as not done

Runus01 committed:
Add support for tracking ToDos, Events, and Deadlines and enable the deletion of tasks.

Runus01 committed:
Add automated text UI testing and handle errors in user input

leroychiu20 commented on src/main/java/TaskManager.java:
```diff
@@ -0,0 +1,124 @@
+ */
+/*
+public class TaskManager {
+    private static final List<Task> THINGS_TO_DO = new ArrayList<>();
```
im not sure if this variable should be declared as final and written in caps, since it is modified in the later part.

leroychiu20 commented on src/main/java/Event.java:
```diff
@@ -0,0 +1,19 @@
+ * Represents a task that is an event.
+ * @author Aaron
+ */
+public class Event extends Task {
+    protected String from;
+    protected String to;
```
you could consider changing 'from' to something like startDate, and 'to' to something like endDate, so that it is clearer and less ambiguous.

leroychiu20 commented on src/main/java/Deadline.java:
```diff
@@ -0,0 +1,17 @@
+ * @author Aaron
+ */
+public class Deadline extends Task {
+    protected String by;
```
you can consider changing 'by' to something like doByDate so that it is clearer for the reader.

leroychiu20 commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,30 @@
+        return isDone ? "X" : " "; // mark a done task with 'X'
+    }
+
+    public void done() {
```
done() isn't really a verb, so you could consider changing it to mark() instead

leroychiu20 commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,30 @@
+        this.isDone = true;
+    }
+
+    public void notDone() {
```
similarly, you may want to consider changing notDone() to unmark()

leroychiu20 reviewed (commented):
LGTM. However, I noted a few areas where naming could be changed.

Runus01 committed:
Save the tasks in an Elsa.txt data file automatically whenever the task list changes and load the saved data to populate the list of tasks when the chatbot starts up.

Runus01 committed:
Merge branch 'branch-Level-7'

dwsc37 commented on src/main/java/Elsa.java:
```diff
@@ -0,0 +1,307 @@
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Scanner;
```
Good organisation of import statements to clearly show packages imported!

dwsc37 commented on src/main/java/Elsa.java:
```diff
@@ -0,0 +1,307 @@
+
+    /**
+     * This method updates the THINGS_TO_DO arrayList to match the data in the Elsa.txt file.
+     */
```
I think it would be good for the JavaDocs to include `@param` and `@return` for functions with parameters/return values.

dwsc37 commented on src/main/java/Elsa.java:
```diff
@@ -0,0 +1,307 @@
+                return new Event(description, isDone, from, to);
+
+            default:
+                throw new IllegalArgumentException("Invalid task type: " + taskType);
```
`case` blocks should have the same indentation as `switch` statement (i.e. everything should be shifted 4 spaces left)

dwsc37 commented on src/main/java/Elsa.java:
```diff
@@ -0,0 +1,307 @@
+        tasks.add(newTodo);
+        addLine();
+        System.out.println("Alright, I've added this task:\n  " + tasks.get(tasks.size() - 1) + "\nWe have " +
+                           tasks.size() + " tasks in our list now.");
```
When breaking a long line into 2 lines, the second line should only be indented 8 spaces

dwsc37 commented on src/main/java/Elsa.java:
```diff
@@ -0,0 +1,307 @@
+        tasks.add(newDeadline);
+        addLine();
+        System.out.println("Alright, I've added this task:\n  " + tasks.get(tasks.size() - 1) + "\nWe have " +
+                           tasks.size() + " tasks in our list now.");
```
Ditto

dwsc37 commented on src/main/java/Elsa.java:
```diff
@@ -0,0 +1,307 @@
+            // Informs the user that the task has been deleted
+            addLine();
+            System.out.println("Okay, I've removed this task:\n  " + tasks.get(taskIndex).toString() + "\nWe have " +
+                               (tasks.size() - 1) + " tasks in our list now.");
```
Ditto

dwsc37 commented on src/main/java/Elsa.java:
```diff
@@ -0,0 +1,307 @@
+        tasks.add(newEvent);
+        addLine();
+        System.out.println("Alright, I've added this task:\n  " + tasks.get(tasks.size() - 1) + "\nWe have " +
+                           tasks.size() + " tasks in our list now.");
```
Ditto

dwsc37 commented on src/main/java/Elsa.java:
```diff
@@ -0,0 +1,307 @@
+     * This method unmarks any task by removing any 'X'.
+     */
+    private static void unmarkTask(List<Task> tasks, String userInput) {
+        // Finds and marks the task the user is referring to as not done
+        int taskIndex = Integer.parseInt(userInput.substring(7)) - 1;
+        if (taskIndex >= 0 && taskIndex < tasks.size()) {
+            tasks.get(taskIndex).notDone();
+            // Informs the user that the task has been marked as not done
```
Good use of comments to explain the logic!

dwsc37 reviewed (commented):
Code is very readable, and largely adheres to the coding standard, just some small nits to fix

Runus01 committed:
Add dates and times for tasks with a deadline

Runus01 committed:
Merge branch 'branch-Level-8'

Runus01 committed:
Add Ui, Storage, Parser, TaskList and Command classes for more OOP

Runus01 committed:
Organise the classes into suitable java packages

Runus01 committed:
Merge branch 'add-gradle-support'

Runus01 committed:
Add Gradle to automate build tasks of the project

Runus01 committed:
Add JUnit tests to test the behaviour of Todo.java and Parser.java

Runus01 committed:
Add the capability to package the app as an executable JAR file

Runus01 committed:
Add JavaDoc comments to the code

Runus01 committed:
Merge branch 'branch-A-JavaDoc'

Runus01 committed:
Update code to comply with coding standard

Runus01 committed:
Merge branch 'branch-A-CodingStandard'

Runus01 committed:
Add the ability to find tasks in the task list

Runus01 committed:
Merge branch 'branch-Level-9'

Runus01 committed:
Amend comment for main method in Elsa.java and add Checkstyle to Gradle

Runus01 committed:
Add Checkstyle configuration to detect coding style violations

Runus01 committed:
Merge branch 'branch-A-CheckStyle'

Runus01 committed:
Add JavaFX-based GUI for the chatbot

Runus01 committed:
Merge branch 'branch-Level-10'

Runus01 committed:
Update .gitignore to exclude test-ui-test folder and ACTUAL.TXT

Runus01 committed:
Add assertions, reduce countdown timer, and display welcome message

Assertions are missing in parts of the code, leading to potential
runtime issues when assumptions in the logic are violated.
The countdown timer for exiting after a 'bye' input is set to 5 seconds,
which delays the application shutdown unnecessarily. The GUI does not
display a welcome message from Elsa when the application starts,
missing an opportunity to enhance user experience.

Assertions ensure that assumptions are explicitly validated, preventing
runtime errors. Reducing the countdown timer from 5 to 3 seconds
shortens the wait time for a smoother exit. Adding a welcome message
from Elsa in the GUI provides immediate feedback when the application
starts, improving usability.

Let's add assertions to catch potential errors and safeguard the code
against unexpected input. Reducing the countdown timer makes the
shutdown process more responsive. Updating the GUI with a welcome
message enhances the first interaction with the application.

Runus01 committed:
Improve input validation for mark, unmark, delete, and event creation

Commands like mark or unmark followed by an invalid number, an empty
string, or no number were causing the program to break.

Allowing these invalid inputs to crash the program results in poor
user experience and system instability. Additionally, tasks like event
and todo allowed empty descriptions, and event tasks failed silently
when the creation format was wrong, leaving users without guidance.

Let's ensure that invalid inputs for mark, unmark, and delete are
properly handled:
* Invalid numbers or no numbers following these commands will no longer
crash the program
* Negative numbers and non-numbers will also be caught and will not
cause the program to break

Let's also enforce input validation for event and todo tasks:
* Empty descriptions will now be rejected for both event and todo tasks
* Incorrect formats when creating event tasks will result in clear
guidance to the user about the correct input format, preventing silent
failures

These changes improve the code quality by making the system more robust
and user-friendly. Input validation and error handling prevent crashes
and improve usability.

Runus01 committed:
Add duplicate task recognition and handling

Tasks are added to the list without checking for duplicates, allowing
the same task to be added multiple times.

This leads to confusion and clutter in the task list, making it harder
for users to manage their tasks effectively.

Let's implement a check for duplicate tasks when adding new tasks.
If a task of the same task type as the new task and with the same
description already exists in the list, throw an ElsaException to
inform the user. This ensures that each task is unique and improves
overall task management.

Using this approach maintains data integrity and enhances user
experience by preventing the addition of redundant tasks.

Runus01 committed:
Merge pull request #2 from Runus01/branch-A-Assertions

Add assertions, reduce countdown timer, and display welcome message

Runus01 committed:
Merge branch 'branch-A-Assertions'

Runus01 committed:
Merge branch 'master' into branch-A-CodeQuality

Runus01 committed:
Merge branch 'master' into branch-BCD-Extension

# Conflicts:
#	src/main/java/elsa/task/TaskList.java

Runus01 committed:
Merge pull request #3 from Runus01/branch-A-CodeQuality

Improve input validation for mark, unmark, delete, and event creation

Runus01 committed:
Merge branch 'branch-A-CodeQuality'

Runus01 committed:
Merge branch 'master' into branch-BCD-Extension

# Conflicts:
#	src/main/java/elsa/task/TaskList.java

Runus01 committed:
Merge pull request #4 from Runus01/branch-BCD-Extension

Add duplicate task recognition and handling

Runus01 committed:
Merge branch 'branch-BCD-Extension'

Runus01 committed:
Add a title to the GUI window and improve the readability of the Parser

Runus01 committed:
Add a user guide to the project

Runus01 committed:
Add JAR configuration to prepare the product for release
